Support for Flamegraph

zemse opened the pull request:
## Motivation

This PR closes #776. Old PR #8315.

## Solution

Adds a `flamegraph` and `flamechart` flag to the test subcommand. Using `--match-test` and/or `--match-contract` just a single test case must be run.

Example:

```sh
# sorts functions by name and merges them
forge test --match-test test_hash_var_len_2 --flamegraph

# does not merge functions and prints them in execution order, basically visualizes the trace chart
forge test --match-test test_hash_var_len_2 --flamechart 
```

This will execute just the `test_hash_1_a_sol` test and generate a `flamegraph.svg` file in `cache` dir and will also try to open this file in the default program.

klkvr commented on Cargo.toml:
```diff
@@ -166,7 +166,9 @@ solang-parser = "=0.3.3"
-revm-inspectors = { version = "0.5", features = ["serde"] }
+# revm-inspectors = { version = "0.5", features = ["serde"] }
+# TODO
+revm-inspectors = { path = "/Users/sohamzemse/Workspace/opensource/external/paradigm/revm-inspectors", features = ["serde"] }
```
could you please turn this into a patch to https://github.com/paradigmxyz/revm-inspectors so that CI works and make this PR into a draft so we don't merge it accidentally?

klkvr commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -358,11 +369,54 @@ impl TestArgs {
+            let mut file = std::fs::File::open(file_name).unwrap();
+            file.read_to_string(&mut buf).expect("failed to read flamegraph file");
+            let buf = buf.replace("samples", "gas");
+            std::fs::write(file_name, buf).expect("failed to write flamegraph file");
```
`.wrap_err(...)?`

klkvr replied:
Let's also add `opener::open(file_name)` here, I think it makes sense to show flamegraph right away, and maybe we should save it to `/tmp` in that case

klkvr commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -358,11 +369,54 @@ impl TestArgs {
+            let mut options = inferno::flamegraph::Options::default();
+            options.flame_chart = true;
+
+            inferno::flamegraph::from_lines(&mut options, fst.iter().map(|s| s.as_str()), writer)
```
Instead of writing to file, reading it and writing again, you can use `Vec<u8>` or any other in-memory type which implements `Write`

klkvr reviewed (changes requested):
Thank you for this! Left several comments

Wondering if `--flamegraph` should automatically enable `--decode-internal`? I think it doesn't hurt but would make flamegraphs more informative by default

It seems that current folded stack trace impl is a bit incorrect. Not yet sure what's the issue, but for me the flamegraph for the following test does not look correct:
```solidity
contract Contract {
    uint256 number;

    function run() public {
        for (uint i = 0; i < 10; i++) {
            number += 1;
        }
    }
}

contract TestContract {
    function test() public {
        Contract c = new Contract();
        c.run();
    }
}
```

zemse commented:
Sorry for the delay. Thanks for the review! 

I have made the changes and rebased the branch. Please re-review.

klkvr commented on crates/evm/traces/src/folded_stack_trace/builder.rs:
```diff
@@ -0,0 +1,157 @@
+#[derive(Debug, Clone, Default)]
+pub(crate) struct FoldedStackTraceBuilder {
+    traces: Vec<(Vec<String>, i64)>,
+    exits: Option<u64>,
```
isn't `None` equivalent to 0 here?

zemse replied:
Yes, my bad, updated this now

klkvr reviewed (changes requested):
lgtm, tested locally and flamegraphs are looking good, UX is smooth too!

let's merge `FoldedStackTraceBuilder` into signle file as it's not that large, and would also appreciate some docs on what it's doing (especially on the struct contents and how those are manipulated)

zemse commented on crates/evm/traces/src/folded_stack_trace/mod.rs:
I've combined the files. Kept `EvmFoldedStackTraceBuilder` separate from the raw `FoldedStackTraceBuilder` for abstracting the logic with tests.

zemse commented:
Thanks for checking it out! I've resolved the review comments, added some comments, and rebased the branch.

klkvr commented on crates/evm/traces/src/folded_stack_trace.rs:
```diff
@@ -0,0 +1,301 @@
+    }
+
+    /// Subtracts gas consumed by the children function calls from the parent function calls.
+    fn subtract_children(&mut self) {
```
couldn't this always handled in `enter`?

zemse replied:
Purposefully kept it for manual debugging in future because final FST is not very human-readable. That's why there's an intermediate check in the test cases before and after subtraction, for FST module to be self-explaining. But if it seem to complicate then I can move the logic to `enter`.

klkvr reviewed (commented):
overall lgtm, have a question re merging of entires:
```rust 
let mut trace = super::FoldedStackTraceBuilder::default();
trace.enter("top".to_string(), 500);
trace.enter("child_a".to_string(), 100);
trace.exit();
trace.enter("child_a".to_string(), 200);
trace.exit();

assert_eq!(
    trace.build_without_subtraction(),
    vec![
        "top 500", //
        "top;child_a 100",
        "top;child_a 200",
    ]
);
```

shouldn't `child_a` entries end up merged here?

iFrostizz commented:
Always found that the `--debug <TEST_NAME>` was a bit broken and conflicting with the current way of selecting tests.
We already have `--matching-test` so adding a `--flamegraph <TEST_NAME>` adds some new handling while you could just return an error if the matching failed to select only 1 test. What we may want though is a way to select a single test efficiently, without having to combine the `--matching-contract` and `--matching-test` in the case of duplicated contract name. Cargo does it well because it has namespaces by default, but could we consider having something like 
```sh
forge test invariants::VaultTest::testDepositTokens --flamegraph
```

iFrostizz commented:
Congrats on this btw!!!

zemse commented:
> shouldn't `child_a` entries end up merged here?

@klkvr Currently, I was generating a flamechart instead. When it's flamegraph then the inferno lib sorts the lines to render with merged functions.

@iFrostizz thanks for feedback, it makes sense to use general tests matching logic.

I have changed the `--flamegraph` to be a boolean flag and use tests using the general test matching API. Also added `--flamechart` which is a variation. So the API looks like

```sh
# sorts functions by name and merges them
forge test --match-test test_hash_var_len_2 --flamegraph

# does not merge functions and prints them in execution order, basically visualizes the trace chart
forge test --match-test test_hash_var_len_2 --flamechart 
```

Please try and would love feedback!

klkvr commented on crates/evm/traces/src/folded_stack_trace.rs:
```diff
@@ -0,0 +1,301 @@
+    /// Exits all the previous internal calls that should end before starting step_idx.
+    fn exit_previous_steps(&mut self, step_exits: &mut Vec<usize>, step_idx: usize) {
+        let initial_length = step_exits.len();
+        step_exits.retain(|&number| number >= step_idx);
```
```suggestion
        step_exits.retain(|&number| number > step_idx);
```

was able to hit this edge case locally (call started right after the terminating step for internal call)

klkvr reviewed (approved):
LGTM, works great for me locally!

left a couple small comments, should be quick fixes

klkvr commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -364,9 +382,60 @@ impl TestArgs {
+                .wrap_err("failed to write svg")?;
+
+            // Open SVG in default program.
+            opener::open(&file_name).wrap_err("failed to open flamegraph")?;
```
Let's handle the error gracefully here by saving flamegraph into `cache/{filename}` to allow user to still take a look at it even if `opener` failed to find a way to open it.

zemse replied:
I've pushed a commit to address this. Pls have a loook

klkvr commented on crates/evm/traces/src/folded_stack_trace.rs:
```diff
@@ -0,0 +1,301 @@
+            let TraceEntry { names, gas } = &right[0];
+            if names.len() > 1 {
+                let parent_trace_to_match = &names[..names.len() - 1];
+                for parent in left {
```
```suggestion
                for parent in left.iter_mut().rev() {
```

this will be more correct as if we already walked the same call path earlier, we'll subtract from the wrong node

klkvr commented:
Note: profiling gas is a bit different vs traditional stack sampling, so there are some bugs I'm observing rn which we can look into in follow-ups. Mostly its related to refunds which result in negative cost code paths or children spending more gas than parents

zerosnacks commented on crates/forge/Cargo.toml:
```diff
@@ -78,6 +78,7 @@ dialoguer = { version = "0.11", default-features = false }
 dunce.workspace = true
 futures.workspace = true
 indicatif = "0.17"
+inferno = "0.11.19"
```
It looks like there is an issue with the license of `inferno`, will need to determine if this is compatible with our licenses

https://github.com/jonhoo/inferno/blob/main/LICENSE

DaniPopes replied:
[cargo-flamegraph](https://github.com/flamegraph-rs/flamegraph) uses it and has the same licenses as us

DaniPopes commented on crates/forge/Cargo.toml:
```diff
@@ -78,6 +78,7 @@ dialoguer = { version = "0.11", default-features = false }
 dunce.workspace = true
 futures.workspace = true
 indicatif = "0.17"
+inferno = "0.11.19"
```
Please disable default features to remove unused dependencies

mattsse reviewed (approved):
ty!

pending dep nits

DaniPopes commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -308,14 +316,25 @@ impl TestArgs {
 
+        let should_debug = self.debug.is_some();
+
+        if self.flamegraph && self.flamechart {
```
This check should be a `#[arg(... conflicts_with = "...")]` in the clap args above

DaniPopes commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -308,14 +316,25 @@ impl TestArgs {
         let env = evm_opts.evm_env().await?;
 
+        // Enable internal tracing for more informative flamegraph.
+        if self.flamegraph || self.flamechart {
```
this `graph || chart` can have its own variable/helper method

DaniPopes commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -364,9 +382,62 @@ impl TestArgs {
             "decode-internal",
             self.decode_internal.as_ref().and_then(|v| v.as_ref()),
         )?;
+        // maybe_override_mt("flamegraph", self.flamegraph.as_ref())?;
```
?

DaniPopes commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -364,9 +382,62 @@ impl TestArgs {
-        let outcome = self.run_tests(runner, config, verbosity, &filter, &output).await?;
+        let mut outcome = self.run_tests(runner, config, verbosity, &filter, &output).await?;
+
+        if self.flamegraph || self.flamechart {
+            let (suite_name, (test_name, test_result)) = outcome
+                .results
+                .iter_mut()
+                .find(|(_, r)| !r.test_results.is_empty())
+                .map(|(suite_name, r)| (suite_name, r.test_results.iter_mut().next().unwrap()))
+                .unwrap();
```
should unify the "get the only test result" logic with `debug` below

DaniPopes commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -81,6 +81,14 @@ pub struct TestArgs {
     #[arg(long, value_name = "TEST_FUNCTION")]
     debug: Option<Regex>,
 
+    /// Generate flamegraph if a single test is executed.
```
```suggestion
    /// Generate a flamegraph for a single test. Implies `--decode-internal`.
```

Please also apply to `flamechart`

DaniPopes commented on crates/forge/bin/cmd/test/mod.rs:
```diff
@@ -81,6 +81,14 @@ pub struct TestArgs {
+    flamegraph: bool,
+
+    /// Generate flamechart if a single test is executed.
+    #[arg(long, env = "FLAMECHART")]
```
I don't think `env` option is needed

gakonst commented:
I cloned the PR and ran it against Solady, but something seems to not be matching? https://github.com/Vectorized/solady

```
 /Users/georgios/.foundry/foundry-rs/foundry/target/debug/forge t --flamegraph testSafeTransferFromToERC721RecipientWithData
No tests match the provided pattern:
        match-path: `testSafeTransferFromToERC721RecipientWithData`
Error:
No tests to run
```

when there's clearly such a test 

```
 rg testSafeTransferFromToERC721RecipientWithData
test/ERC721.t.sol
776:    function testSafeTransferFromToERC721RecipientWithData(uint256 id, bytes memory data) public {
```

klkvr commented:
@gakonst this should be `forge t --flamegraph --mt testSafeTransferFromToERC721RecipientWithData`

`--flamegraph` does not accept test function arg, thus function name was parsed as cli arg (equivalent to `--match-path`)

zemse commented:
I've updated the PR description to reflect the current code. Sorry for the inconvenience.
